website/addons/onedrive/model.py

```diff
@@ -0,0 +1,240 @@
+# -*- coding: utf-8 -*-
+import logging
+
+from datetime import datetime
+
+from modularodm import fields
+
+from framework.auth import Auth
+
+from website.addons.base import exceptions
+from website.addons.base import AddonOAuthUserSettingsBase, AddonOAuthNodeSettingsBase
+from website.addons.base import StorageAddonBase
+
+from website.addons.onedrive import settings
+from website.addons.onedrive.utils import OneDriveNodeLogger
+from website.addons.onedrive.serializer import OneDriveSerializer
+from website.addons.onedrive.client import OneDriveAuthClient
+from website.addons.onedrive.client import OneDriveClient
+
+from website.oauth.models import ExternalProvider
+
+logger = logging.getLogger(__name__)
+
+logging.getLogger('onedrive1').setLevel(logging.WARNING)
+
+
+class OneDrive(ExternalProvider):
+    name = 'onedrive'
+    short_name = 'onedrive'
+
+    client_id = settings.ONEDRIVE_KEY
+    client_secret = settings.ONEDRIVE_SECRET
+
+    auth_url_base = settings.ONEDRIVE_OAUTH_AUTH_ENDPOINT
+    callback_url = settings.ONEDRIVE_OAUTH_TOKEN_ENDPOINT
+    auto_refresh_url = settings.ONEDRIVE_OAUTH_TOKEN_ENDPOINT
+    default_scopes = ['wl.basic wl.signin onedrive.readwrite wl.offline_access']
+
+    _auth_client = OneDriveAuthClient()
+    _drive_client = OneDriveClient()
+
+    def handle_callback(self, response):
+        """View called when the Oauth flow is completed. Adds a new OneDriveUserSettings
+        record to the user and saves the user's access token and account info.
+        """
+        userInfo = self._auth_client.user_info(response['access_token'])
+        #  userInfo = userInfoRequest.json()
+        logger.debug("userInfo:: %s", repr(userInfo))
+
+        return {
+            'provider_id': userInfo['id'],
+            'display_name': userInfo['name'],
+            'profile_url': userInfo['link']
+        }
+
+    def _refresh_token(self, access_token, refresh_token):
```

Comment:
This and the two other `refresh`-related functions in this class are likely superfluous, with `auto_refresh_url` defined. See `website/oauth/models/__init__.py:ExternalProvider.refresh_oauth_key` -- it does all of this work. This would also obviate `OneDriveAuthClient.refresh`; `.user_info` could be moved into `OneDriveClient`.